// innerscope/Dist/router.js
export default class Router {
    routeMap;
    pageRoutes;
    rootPage;
    routerMode;
    currentIndex = 0;
    constructor(rootPage, configuration) {
        this.routeMap = new Map();
        this.rootPage = rootPage;
        this.routerMode = configuration.mode;
        this.pageRoutes = configuration.routes;
        this.pageRoutes?.forEach((route) => {
            this.AddRoute(route.path, route.component);
        });
        // Initialize the current index based on the initial state
        this.currentIndex = window.history.state?.index || 0;
        // assign according to prefered router mode
        if (this.routerMode === "hash") {
            var route = window.location.hash.slice(1);
            window.onhashchange = (event) => {
                this.hash_change_handler(route);
            };
        }
        else if (this.routerMode === "history") {
            var route = window.location.pathname;
            window.onpopstate = (event) => {
                const newIndex = event.state?.index || 0;
                this.currentIndex = newIndex;
                this.popstate_handler(route, event);
            };
        }
    }
    AddRoute(route, Function) {
        this.routeMap.set(route, Function);
    }
    Open(path) {
        if (!this.does_route_exist(path)) {
            path = "/404";
        }
        if (this.routerMode == "hash") {
            this.hash_change_handler(path);
        }
        else
            this.popstate_handler(path);
        // console.log("Opening Page: " + path);
    }
    does_route_exist(path) {
        return this.routeMap.has(path);
    }
    hash_change_handler(route) { }
    popstate_handler(route, event) {
        // console.log("PopStateHandler Recieving: ", route);
        const component = this.routeMap.get(route);
        document.body.innerHTML = "";
        component ? component() : console.error();
        const newIndex = this.currentIndex + 1;
        this.currentIndex = newIndex;
        // console.log(`Navigated to ${route}`);
        history.pushState({ index: newIndex }, "", route);
    }
}
